From 802a7b483468e7fc22996c7f644efd8730577d3d Mon Sep 17 00:00:00 2001
From: Ariadne Conill <ariadne@ariadne.space>
Date: Wed, 15 Oct 2025 15:15:52 -0700
Subject: [PATCH] commit: add support for system-provided commit hooks

presently, packages which use commit hooks in order to provide package-specific
functionality install them into /etc/apk/commit_hooks.d, which is intended for
user-configured commit hooks only.

ref 349c61c9612a add support for pre and post commit hooks
---
 doc/apk.8.scd | 3 ++-
 src/commit.c  | 7 +++++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/doc/apk.8.scd b/doc/apk.8.scd
index bd15fe7..47e4772 100644
--- a/doc/apk.8.scd
+++ b/doc/apk.8.scd
@@ -402,7 +402,8 @@ timeout 120
 	This is expected to be a symlink to directory what apk will use
 	as package cache. See also *apk-cache*(5) and *apk-cache*(8).
 
-*/etc/apk/commit_hooks.d/\**
+*/etc/apk/commit_hooks.d/\**++
+*/lib/apk/commit_hooks.d/\**
 	Hook scripts which are executed before anything has been written to the
 	filesystem and after all the changes have been commited. The script
 	executed gets as an argument the stage name (*pre-commit* or
diff --git a/src/commit.c b/src/commit.c
index e36e167..2f86791 100644
--- a/src/commit.c
+++ b/src/commit.c
@@ -271,8 +271,11 @@ static int run_commit_hook(void *ctx, int dirfd, const char *file)
 static int run_commit_hooks(struct apk_database *db, int type)
 {
 	struct apk_commit_hook hook = { .db = db, .type = type };
-	return apk_dir_foreach_file(openat(db->root_fd, "etc/apk/commit_hooks.d", O_DIRECTORY | O_RDONLY | O_CLOEXEC),
-				    run_commit_hook, &hook);
+	return apk_dir_foreach_config_file(db->root_fd,
+	    	run_commit_hook, &hook, NULL,
+		"etc/apk/commit_hooks.d",
+		"lib/apk/commit_hooks.d",
+		NULL);
 }
 
 static int calc_precision(unsigned int num)
-- 
2.51.0

